### DISCUSSION TITLE
How to use static linked C++ library via FFI

### DISCUSSION THREAD

**Discussion by noname0310** [wasm-bindgen/wasm-bindgen#3886] [Q&A] [answered]
### Summary

I want to use [bullet physics](https://github.com/bulletphysics/bullet3), a physics engine library written in c++ via FFI, in wasm rust.

### Additional Details

To do this, I found out that I need to build the c++ library to nostd with cc.

So I added includes to provide std of c++ from rust to ffi
then I tried build btScalar.h in bullet physics with the build code below
```rust
// build.rs
fn main() {
    cc::Build::new()
        .archiver("llvm-ar")
        .cpp_link_stdlib(None)
        .cpp(true)
        .flag("-xc++")
        .includes([
            "cpp_wasm_std"
        ])
        .files([
            "cpp_src/lib.h",
            "cpp_src/LinearMath/btScalar.h"
        ])
        .compile("bullet");

    println!("cargo:rerun-if-changed=cpp_wasm_std");
    println!("cargo:rerun-if-changed=cpp_src/lib.h");
    println!("cargo:rerun-if-changed=cpp_src/LinearMath/btScalar.h");
}
```

The build was successful but wasm-bindgen is panicking. I don't know if this is a bug:
```sh
   Compiling cpp-wasm-ffi v0.1.0 (D:\noname\Projects\GitHub\rust-wasm-cpp-ffi\src\wasm_src)
    Finished dev [unoptimized + debuginfo] target(s) in 0.14s
[INFO]: Installing wasm-bindgen...
thread 'main' panicked at crates\wasm-interpreter\src\lib.rs:311:32:
index out of bounds: the len is 1024 but the index is 534773760
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Error: Running the wasm-bindgen CLI
Caused by: Running the wasm-bindgen CLI
Caused by: failed to execute `wasm-bindgen`: exited with exit code: 101
```

Here is a repository that reproduces the issue:
https://github.com/noname0310/rust-wasm-cpp-ffi

**Comment by daxpedda** [comment]
I don't know if this is actually the bug you are hitting, but Rust `extern "C"` ABI is not compatible with ABIs from other compilers for the Wasm target.

See https://github.com/rustwasm/wasm-bindgen/issues/3454.

**Reply by noname0310** [reply]
It worked just fine in a simpler scenario. This is the worked version of the source.
https://github.com/noname0310/rust-wasm-cpp-ffi/tree/6b7eb3cc5216dc6fbe5ef1e82b605495550ee6ff

**Reply by daxpedda** [reply]
Not all parts of the ABI are incompatible, AFAICS in your example you are passing some integers only, which should be fine indeed.

See https://github.com/rust-lang/rust/issues/71871 for more information.

**Reply by noname0310** [reply]
According to that, there would have been a linking error. 

What wasm-bindgen panics about is nonsense...

**Reply by daxpedda** [reply]
> According to that, there would have been a linking error.

Good point.

> What wasm-bindgen panics about is nonsense...

The interpreter seems to have some issue parsing something specific.
If you can narrow down what it is that would be helpful.

Old versions of `wasm-bindgen` have also shown this error when there was a version mismatch between the Rust dependency and the CLI. But it doesn't *seem* like that's the issue you are having.

**Reply by noname0310** [reply]
I'll try narrow down the cause of the problem and ask again.

**Comment by Liamolucko** [comment, marked as answer]
It looks like LLVM's generating `__wasm_call_ctors` calls again:

```wat
(func $__wbindgen_describe_init.command_export (;1496;) (type 14)
  call $__wasm_call_ctors
  call $__wbindgen_describe_init
)
```

This has come up before, but only in the context of compiling to WASI (#2969). Previously I'd assumed that `__wasm_call_ctors` was just something WASI-specific, but now I'm thinking it might just get generated automatically whenever there's code that needs global initialisation, which included all WASI code. (This is all speculation, I haven't actually double-checked this yet.)

It seems like #2673 may actually have been the same issue since the OP there also mentioned C++, but it never went anywhere.

After some quick googling for what `__wasm_call_ctors` is in the first place, I found https://github.com/WebAssembly/WASI/issues/471; it seems like a workaround may be to generate these calls manually on all the entrypoints to Wasm, which signals to LLVM that it doesn't have to do it for us and so it won't do it on the descriptors where they aren't wanted.

Also note that this only happens in debug builds, it goes away in release mode. My guess as to why is that the compiler can see that the global initialisation is just setting some memory to a fixed value, and so replaces it with a data segment instead.

**Reply by Liamolucko** [reply]
Something to double-check before implementing that workaround is whether or not `__wasm_call_ctors` is meant to be called more than once. We might need to have some kind of global flag that gets set once it's been called to stop it being called again; I have zero idea, and I don't have time to research it right now.

**Reply by noname0310** [reply]
Thank you very much for your advice, I externed __wasm_call_ctors and then called it directly and it worked fine.

I also confirmed that the issue only occurs in debug mode.

https://github.com/noname0310/rust-wasm-cpp-ffi/commit/b44cc94b199a5e4b443e54f0181e866e95828eb7
